File: Strings/4/myOptimized.js
var backspaceCompare = function (S, T) {
  let p1 = S.length - 1, p2 = T.length -1, skipS = 0, skipT = 0;
  while (p1 >= 0 || p2 >=0) {
    while(p1 >= 0) {
      if (S[p1] === '#') {
        p1--;
        skipS++;
      }
      else if (skipS > 0) {
        p1--;
        skipS--;
      }
      else {
        break;
      }
    }

    while(p2 >= 0) {
      if (T[p2] === '#') {
        p2--;
        skipT++;
      }
      else if (skipT > 0) {
        p2--;
        skipT--;
      }
      else {
        break;
      }
    }

    if (p1 >=0 && p2 >=0 && S[p1] !== T[p2]) return false;
    if ((p1 >= 0) != (p2 >=0)) return false;
    p1--;
    p2--;
  }
  return true;
};

console.log(backspaceCompare((S = 'abab##'), (T = 'ab')));

File: Strings/4/myOptimized.py
def backspaceCompare(S, T):
    p1, p2 = len(S)-1, len(T)-1
    skipA, skipB = 0, 0
    while p1 >= 0 or p2 >= 0:
        while p1 >= 0:
            if S[p1] == '#':
                skipA += 1
                p1 -= 1
            elif skipA > 0:
                skipA -= 1
                p1 -= 1
            else:
                break
        while p2 >= 0:
            if T[p2] == '#':
                skipB += 1
                p2 -= 1
            elif skipB > 0:
                skipB -= 1
                p2 -= 1
            else:
                break
        if p1 >= 0 and p2 >= 0 and S[p1] != T[p2]:
            return False
        if (p1 >= 0) != (p2 >= 0):
            return False
        p1 -= 1
        p2 -= 1
    return True

print(backspaceCompare(S="bxj##tw", T="bxo#j##tw"))

'''
#ac#b  b
abab###  a
###a#   ""
###    ""
##aa#   a
'''


